### ISSUE TITLE
Invisible creatures

### ISSUE THREAD

**Issue by edf42001** [evolvio/evolv.io#52]
There is a line of code in the reproduce method that says:
 double newPY = random(-0.01, 0.01); //To avoid landing directly on parents, resulting in division by 0)

I believe this still sometimes happens, and creates invisible creatures and weird land, as shown in the images below

Has this been fixed, or, how about this:
 while(newPY == 0){
          newPY = random(-0.01, 0.01);
        }

![invisible creatures 1](https://cloud.githubusercontent.com/assets/19157596/19789650/7b8c8d7a-9c7f-11e6-98ee-3ef8ad0f5957.PNG)
![invisible creatures 2](https://cloud.githubusercontent.com/assets/19157596/19789651/7ed6473c-9c7f-11e6-8483-95d13773e7a6.PNG)
![weird land](https://cloud.githubusercontent.com/assets/19157596/19789883/4a8d1bb6-9c81-11e6-982f-7a43b059297e.PNG)

Also, how do you get the code to have that grey box surrounding it?

**Comment by SirSKillz** [comment]
Are you sure that this is the issue with the code? The likelihood of you getting 0 is very, very small, but your creatures seem to have all been born near the same time, so it seems like it would have to be something other than that small chance. Can you think of anything else it could be?

**Comment by edf42001** [comment]
A few thoughts/observations:

First of all, a creature that goes onto a weird tile sometimes, but not always, turns invisible. That could explain why they all seem to be born around the same time. 

I was messing with something to make it so the creatures can't fight after they give birth for a while (to protect the baby), and around that time this seemed to happen more often. It is possible it has something to do with that, but I remember it happening before. 

Also, this is definitely ONE way of making this happen, because if you just set newPY (and PX) to be 0 it happens much more often. 

Assuming there are not other ways of causing this problem (there may be), the actual issue is when you divide by distance (which is 0) in calculating the collision force, because adding a small value to distance gets rid of the problem if newPX and newPY are set to 0.

And is it really such a small chance that it won't happen for 50,000 creatures? 

Is it possible it spawns really close to the center of the creature and is pushed so the distance is 0?

To sum up: I know this is one way of causing the problem, but I am not sure it is the only one or the one actually causing it during simulations.

**Comment by SirSKillz** [comment]
"\uFFFD" is the unicode character 'Replacement Character' for java, C and C++, as far as I know, so that is more likely. That is going to be very difficult to find... Are you using the master for these findings, or is it with your addition? I only ask this because I have never seen this issue on my simulation. If it is on the master, do you mind if I assign this issue to you?

**Comment by edf42001** [comment]
What if we fix the effects instead of the cause, because we don't know the cause and it happens so rarely?

In tile class, iterate() method: 
```
if(Double.isNan(foodLevel)){ //fix bug
foodLevel=0;
}
```
In Board class
```
if (me.getRadius() < MINIMUM_SURVIVABLE_SIZE || Double.isNaN(me.getRadius())) {
        me.returnToEarth();
 }
```

Or would that be bad coding practice?

**Comment by mbertheau** [comment]
Since so far this bug doesn't seem to surface often, I suggest we take the time to understand where it's coming from and fix the root cause :)

**Comment by SirSKillz** [comment]
My thoughts exactly. It's nice to cover it up, but it is much better for our foundation if we eliminate as many bugs as we can. @edf42001 do you mind if I set you in charge of finding this bug?

**Comment by edf42001** [comment]
Sure, but it might take a while, and I am not 100% sure I know what I am doing.

**Comment by SirSKillz** [comment]
That's fine, as long as it gets done eventually. For now, I'm going to set the deadline as Nov. 19th, just so we can set up a proper goal.

**Comment by edf42001** [comment]
I'm sorry, but I looked into it and I have no idea what the cause is. My speculations are:
Something with collisions
Something with creatures moving too fast
The random number gets close to 0 and when dividing, it rounds to infinity

**Comment by kodomazer** [comment]
I haven't had that issue myself.  What is the commit hash of the code you're using?

**Comment by edf42001** [comment]
I downloaded the most recent at the time (a9eed6e86d) but because it happens so rarely (Like 5000ish years) I changed the code in the reproduce method from newPY = random(-0.01, 0.01) to 
newPY = 0.000000001. 
That makes it happen more frequently. It is possible it makes it happen for a different reason.

**Comment by SirSKillz** [comment]
Sorry I haven't gotten back to you for awhile. Life got me for a bit so I didn't have time to come back to this. Thanks for staying on time, even if you didn't manage to find out the cause. I would say try to create an exception for what you think might cause it and then have everything be in a try statement. This will let you find where the issue is occurring and then you can keep narrowing down where it could be until you know it is caused by a specific area. After we find where it happens, it will be easier to repair.

**Comment by SirSKillz** [comment]
Whoops... Misclick.

**Comment by 10carfernandez** [comment]
It looks like the problem is that at some point, the creature's values (hue, mouthHue, px, py, etc.) get set to NaN.

I set a break point in Creature.java around line 322 where it says:

this.evolvioColor.text("X: " + EvolvioColor.nf((float) selectedCreature.px, 0, 2), 10, 350);
and then clicked on an invisible creature. What I saw was that the value for selectedCreature.px, as well as some of the other values, where set to "NaN."

The values set to NaN were the following:
energy
hue
mouthHue
px
py
rotation
vr
vx
vy

From Softbody, I think the reason saturation and brightness don't appear as NaN is because they get set equal to 1 at the end of the reproduce() method in Creature.

What throws me off, though, is that this seems to happen to the parents, and not the babies....

**Comment by NateUniAccount** [comment]
Another instance of this bug from @UnknownEvil 

https://cloud.githubusercontent.com/assets/12143899/24174455/04466c4a-0e67-11e7-9f6a-5b0b73fdcf7c.png
It's only happening in this list (Oldest), Biggest, Youngest, Etc. are unaffected
Selecting them takes you to some unknown location?
